### PR TITLE
Rewrite suggestions for BaseFloatingPicker (experiments)

### LINKED ISSUE #3969: Refactor SuggestionsController and Suggestions with clearer responsibility divide
#### Priorities and help requested (not applicable if asking question):

Are you willing to submit a PR to fix? Yes

Requested priority: Low

Products/sites affected: (if applicable): OWA

#### Describe the issue:

The suggestionsController and suggestions should be separate, but work well together. Possibly make SuggestionsController a react element whose render function only returns `<Suggestions>.` That way it can have a ref to suggestions, which in turn exposes an api for stuff like clicking the "search more" button. The main goal, in my mind, is to keep the "state manager" (SuggestionsController, maybe it should get renamed?) separate from the "view" (Suggestions). That way it's easy to mix and match the two and each one is clean, there isn't a conflating of state and form.

I think the BasePicker (or any component that uses suggestions/suggestionsController) has to deal with way to much knowledge of what the suggestions has. If we could create a clearer divide between the responsibilities between the two we might be able to remove some of the suggestion specific logic in the BasePicker for changing selection inside of Suggestions

Mentioned in #3929

### PR DESCRIPTION
**Pull request by amyngu**
#### Pull request checklist

- [x] Addresses an existing issue: Fixes #3969 
- [x] Include a change request file using `$ npm run change`

#### Description of changes
Instead of using the `Suggestions`, `SuggestionsController` in fabric core, rewrite suggestions to have better responsibility divide
- `Suggestions`: Renders and takes care of selection of ONLY suggestion items
- `SuggestionsControl`: Renders and takes care of selection of suggestions AND customizable header and footers
  - No more passing multiple props for headers and footers
    - consumers will provide an array of ISuggestionHeaderFooterProps that gives a rendering function, an action (if it is to be selectable), and callback for whether it should be shown
- `SuggestionStore`: Data layer of the suggestions, updates for suggestions go through the store and the store is passed to the component (either Suggestions or SuggestionsControl) to render. (Based off of the `SuggestionController` in fabric core, but has no information on selection, only suggestions)
- Now, consumers of Suggestions or SuggestionsControl will not have to know as much about the state of the selected selections to refocus on the suggestions or the header/footer buttons (see removal around onKeyDown in `BaseFloatingPicker`

### CODE REVIEW

**Comment by amyngu** [comment]
This is definitely a major change for experiments (BaseFloatingPicker) but when I ran npm run change, it only gave me minor and patch options. Can anyone advise on what to do?

File: packages/experiments/src/components/FloatingPicker/BaseFloatingPicker.tsx [hunk 1]
```diff
@@ -367,66 +348,30 @@ export class BaseFloatingPicker<T, P extends IBaseFloatingPickerProps<T>> extend
...
-              this.onSuggestionSelect();
-            }
-          }
         }
         break;
 
       case KeyCodes.down:
-        if (suggestionElement && suggestionElement.tryHandleKeyDown(keyCode, this.suggestionStore.currentIndex)) {
+        if (this.suggestionsControl && this.suggestionsControl.handleKeyDown(keyCode)) {
           ev.preventDefault();
```
Comment (joschect) [inline packages/experiments/src/components/FloatingPicker/BaseFloatingPicker.tsx]: I really like moving all of this into another component.

File: packages/experiments/src/components/FloatingPicker/Suggestions/SuggestionsControl.tsx [hunk 2]
```diff
@@ -0,0 +1,467 @@
...
+
+  protected _forceResolveButton: IButton;
+  protected _searchForMoreButton: IButton;
+  protected _selectedElement: HTMLDivElement;
+  protected _suggestions: Suggestions<T>;
+  private SuggestionsOfProperType: new (props: ISuggestionsProps<T>) => Suggestions<T> =
+  Suggestions as new (props: ISuggestionsProps<T>) => Suggestions<T>;
+
+  constructor(suggestionsProps: ISuggestionsControlProps<T>) {
+    super(suggestionsProps);
```
[the remark this answers is not in the document]
Comment (joschect) [reply packages/experiments/src/components/FloatingPicker/Suggestions/SuggestionsControl.tsx]: You'll also need to reset this._suggestions in componentwillreceiveProps in the event that a new suggestions is passed in.

[the remark this answers is not in the document]
Comment (amyngu) [reply packages/experiments/src/components/FloatingPicker/Suggestions/SuggestionsControl.tsx]: this._suggestions is the suggestions element (Suggestions<T>, not the list of suggestions SuggestionModel<T>[]. It is set in the _renderSuggestions method

File: packages/experiments/src/components/FloatingPicker/BaseFloatingPicker.tsx (hunk 1 above)
Comment (amyngu) [reply packages/experiments/src/components/FloatingPicker/BaseFloatingPicker.tsx]: Yes I think the boundary of responsibility is much cleaner this way

**Comment by dzearing** [comment]
@amyngu we don't follow semver for the experiments package. That's why we created it in the first place; to keep teams experimenting without waiting for major release whenever they want to refine an experiment.

**Pull request merged**